src/main/java/edu/iu/terracotta/service/app/impl/QuestionSubmissionServiceImpl.java

```diff
@@ -304,4 +306,38 @@ public void validateQuestionSubmission(QuestionSubmissionDto questionSubmissionD
             throw new DataServiceException("Error 105: There is invalid data in the request. No question submissions or answer submissions will be updated: " + ex.getMessage());
         }
     }
+
+    @Override
+    public boolean canSubmit(String canvasCourseId, String assignmentId, String canvasUserId, long deploymentId) throws CanvasApiException,
+            IOException {
+        //We find the right deployment:
+        Optional<ToolDeployment> toolDeployment = allRepositories.toolDeploymentRepository.findById(deploymentId);
+        if (toolDeployment.isPresent()) {
+            String litDeploymentId = toolDeployment.get().getLtiDeploymentId();
+            List<PlatformDeployment> platformDeploymentList = allRepositories.platformDeploymentRepository
+                    .findByToolDeployments_LtiDeploymentId(litDeploymentId);
+            if (!platformDeploymentList.isEmpty()) {
+                PlatformDeployment platformDeployment = platformDeploymentList.get(0);
+
+                Optional<AssignmentExtended> assignmentExtended = canvasAPIClient.listAssignment(canvasCourseId,
+                        Integer.parseInt(assignmentId), platformDeployment);
+                List<edu.ksu.canvas.model.assignment.Submission> submissionsList = canvasAPIClient.listSubmissions(Integer.parseInt(assignmentId),
+                        canvasCourseId, platformDeployment);
+
+                Optional<edu.ksu.canvas.model.assignment.Submission> submission = submissionsList.stream().filter(sub -> sub.getUser().getId() == Integer.parseInt(canvasUserId)).findFirst();
+                if (assignmentExtended.isPresent() && submission.isPresent()) {
+                    AssignmentExtended assignment = assignmentExtended.get();
+                    int allowedAttempts = assignment.getAllowedAttempts();
+                    int attempt = 0;
+                    if (submission.get().getAttempt() != null && allowedAttempts >0) {
+                        attempt = submission.get().getAttempt();
+                        if (attempt >= allowedAttempts) {
```

Comment:
So if I understand this, this gets the attempt number of the first submission for the user found in the list of submissions. But don't you want to get the maximum attempt number of all of the user's submissions?  Or alternatively just count the user's submissions?

That is, I assume that if a student has, say, 3 submissions, the API call with return 3 submissions objects for that student, and one of them will have attempt=1, one will have attempt=2 and one will have attempt=3. It's not clear to me that the first submission found for the user would be the third attempt with attempt=3, so seems like you need to take a maximum of them all.

Reply:
@machristie It seems it returns the highest attempt always.